presto-spi/src/main/java/io/prestosql/spi/block/Int128ArrayBlockEncoding.java

```diff
@@ -54,13 +67,34 @@ public Block readBlock(BlockEncodingSerde blockEncodingSerde, SliceInput sliceIn
         boolean[] valueIsNull = decodeNullBits(sliceInput, positionCount).orElse(null);
 
         long[] values = new long[positionCount * 2];
-        for (int position = 0; position < positionCount; position++) {
-            if (valueIsNull == null || !valueIsNull[position]) {
-                values[position * 2] = sliceInput.readLong();
-                values[(position * 2) + 1] = sliceInput.readLong();
+        if (valueIsNull == null) {
+            sliceInput.readBytes(Slices.wrappedLongArray(values));
+        }
+        else {
+            int nonNullPositionCount = sliceInput.readInt();
+            sliceInput.readBytes(Slices.wrappedLongArray(values, 0, nonNullPositionCount * 2));
+            int position = nonNullPositionCount - 1;
+            for (int i = positionCount - 1; i >= 0 && position >= 0; i--) {
+                if (!valueIsNull[i]) {
+                    values[2 * i] = values[2 * position];
```

Comment:
could you try to remove multiplication and simply decrement counters by two? I'm pretty sure removing multiplication would improve perf.

Reply:
This is the kind of thing you’d have to catch with perfasm. Multiplication by 2 might well be folded into a bit shift at which point it’s anyone’s guess whether that’s faster than addition of 2.